students/Pirouz_N/lesson03/slicinglab.py

```diff
@@ -0,0 +1,502 @@
+    if not isinstance(seq, Sequence):
+        raise TypeError('Sequence is not a member of collections.Sequence.')
+    if len(seq) < minlength:
+        raise ValueError('Minimum length of sequence is {}.'.format(minlength))
+
+
+def exchange_first_last(seq):
+    """"This function swaps first element and last element of list.
+
+    Args:
+        seq: Is the sequence passed to the function to swap first and last element of.
+
+    Returns:
+        A shallow copy of sequence where the first and last elements are exchanged.
+
+    Raises:
+        ValueError: If length of sequence is less than one.
+        TypeError: If None is passed instead of sequence.
+            If seq is not of type collection.Sequence.
+    """
+    # Protecting against bad inputs
+    check_sequence(seq, 1)
+
+    # Creating a shallow copy of the sequence.
+    # new_seq = lis(seq) would work the same way as a shallow copy
+    new_seq = seq[:]
+
+    # Sequence length one needs to be handled separately
+    if len(new_seq) == 1:
+        return new_seq
+
+    # Creating memory placeholders for for first and last elements
+    first = new_seq[0: 1]
+    last = new_seq[len(seq)-1: len(seq)]
+
+    # Returning corrected sequence
+    return last + new_seq[1: len(new_seq) - 1] + first
+
+
+def exchange_first_last_deep(seq):
+    """"This function swaps first element and last element of list.
+
+    Args:
+        seq: Is the sequence passed to the function to swap first and last element of.
+
+    Returns:
+        A deep copy of sequence where the first and last elements are exchanged.
+
+    Raises:
+        ValueError: If length of sequence is less than one.
+        TypeError: If None is passed instead of sequence.
+            If seq is not of type collection.Sequence.
+    """
+
+    # Protecting against bad inputs
+    check_sequence(seq, 1)
+
+    # Creating a shallow copy of the sequence
+    new_seq = copy.deepcopy(seq)
+
+    # Sequence length one needs to be handled separately
+    if len(new_seq) == 1:
+        return new_seq
+
+    # Creating memory placeholders for for first and last elements
+    first = new_seq[0: 1]
+    last = new_seq[len(seq)-1: len(seq)]
+
+    # Returning corrected sequence
+    return last + new_seq[1: len(new_seq) - 1] + first
+
+
+def alternate_item_deep(seq):
+    """"This function removes every other item starting with first element.
+
+    Args:
+        seq: Is the sequence passed to the function to remove every other item.
+
+    Returns:
+        A deep copy of sequence where every other item is removed.
+
+    Raises:
+        ValueError: If length of sequence is less than one.
+        TypeError: If None is passed instead of sequence.
+            If seq is not of type collection.Sequence.
+    """
+
+    # Protecting against bad inputs
+    check_sequence(seq, 1)
+
+    # Creating a shallow copy of the sequence
+    new_seq = copy.deepcopy(seq)
+
+    # Retuning alternate series
+    return new_seq[:: 2]
```

Comment:
there is no need for copy; the slice syntax creates a new copy

Reply:
Thank you. I added alternate_item function which is a shallow alternate item copy and uses just slicing.  
I could be wrong, if that is okay with you I will leave alternate_item_deep as is? Which first creates a deep copy then uses alternate slicing to pick alternate items from the deep copy. Please let me know if there is a better way to create an alternate deep copy?

Best Regards
Pirouz Naghavi